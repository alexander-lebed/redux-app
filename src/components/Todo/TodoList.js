// @flow
/* eslint-disable no-shadow, no-param-reassign */
import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Col from 'react-bootstrap/lib/Col';
import Row from 'react-bootstrap/lib/Row';
import type { State, Dispatch } from '../../redux/types';
import { addTodo, updateTodo, toggleTodo, removeTodo } from '../../redux/actions/todo-actions';
import Todo from './Todo';


type Props = {
    currentData: State,
    addTodo: (text: string) => void,
    updateTodo: (id: string, text: string) => void,
    toggleTodo: (id: string) => void,
    removeTodo: (id: string) => void
}

export class TodoList extends React.Component<void, Props, void> {

    onSubmit = (event: Object) => {
        const text = event.target.value;
        const isEnterKey = event.which === 13;
        if (isEnterKey && text.length > 0) {
            this.props.addTodo(text);
            event.target.value = '';
        }
    }

    render() {
        const { currentData, updateTodo, toggleTodo, removeTodo } = this.props;
        const todos = currentData.get('todos');
        return (
            <div className="todo">
                <Row>
                    <Col xs={3} />
                    <Col xs={6}>
                        <input
                            type="text"
                            className='todo__entry'
                            placeholder="Add todo"
                            onKeyDown={this.onSubmit}
                        />
                        <ul className="todo__list">
                            {todos.filter(t => !t.get('isDeleted')).map(t => (
                                <li
                                    key={t.get('id')}
                                    className="todo__item"
                                    onClick={() => toggleTodo(t.get('id'))}
                                >
                                    <Todo
                                        todo={t}
                                        onEdit={(id, text) => updateTodo(id, text)}
                                        onDelete={() => removeTodo(t.get('id'))}
                                    />
                                </li>
                            ))}
                        </ul>
                    </Col>
                    <Col xs={3} />
                </Row>
            </div>
        );
    }
}

export default connect(
    // mapStateToProps
    state => ({ currentData: state.todoData }),
    // mapDispatchToProps
    (dispatch: Dispatch) =>
        bindActionCreators({ addTodo, updateTodo, toggleTodo, removeTodo }, dispatch)
        // return {
        //     addTodo: text => dispatch(addTodo(text)),
        //     toggleTodo: id => dispatch(toggleTodo(id)),
        //     removeTodo: id => dispatch(removeTodo(id))
        // };

)(TodoList);
