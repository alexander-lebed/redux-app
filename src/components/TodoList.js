import React from 'react';
import {connect} from 'react-redux';
import Col from 'react-bootstrap/lib/Col';
import Row from 'react-bootstrap/lib/Row';
import Glyphicon from 'react-bootstrap/lib/Glyphicon';
import {addTodo, toggleTodo, removeTodo} from '../actions/todo-actions';


const TodoList = (props) => {
    const {todos, addTodo, toggleTodo, removeTodo} = props;

    const onSubmit = (event) => {
        let text = event.target.value;
        const isEnterKey = event.which === 13
        if (isEnterKey && text.length > 0) {
            addTodo(text);
            event.target.value = '';
        }
    };

    const toggleClick = id => event => toggleTodo(id);
    const onDelete = id => event => removeTodo(id);

    return (
        <div className="todo">
            <Row>
                <Col xs={3} />
                <Col xs={6}>
                    <input
                        type="text"
                        className='todo__entry'
                        placeholder="Add todo"
                        onKeyDown={onSubmit}
                    />
                    <ul className="todo__list">
                        {todos.filter(t => !t.get('isDeleted')).map(t => (
                            <li
                                key={t.get('id')}
                                className="todo__item"
                                onClick={toggleClick(t.get('id'))}
                            >
                                <Todo todo={t} onDelete={onDelete(t.get('id'))} />
                            </li>
                        ))}
                    </ul>
                </Col>
                <Col xs={3} />
            </Row>
        </div>
    );
};

const Todo = (props) => {
    const {todo, onDelete} = props;
    return <div>
        {todo.get('isDone') ? <strike>{todo.get('text')}</strike> : <span>{todo.get('text')}</span>}
        <Glyphicon
            glyph="remove"
            className="pull-right"
            onClick={onDelete}
        />
    </div>
};

export default connect(
    // mapStateToProps
    (state) => {
        return {
            todos: state.todos
        }
    },
    // mapDispatchToProps
    (dispatch) => {
        return {
            addTodo: text => dispatch(addTodo(text)),
            toggleTodo: id => dispatch(toggleTodo(id)),
            removeTodo: id => dispatch(removeTodo(id))
        };
    }
)(TodoList)