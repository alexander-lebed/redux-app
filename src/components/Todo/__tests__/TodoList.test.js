// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { List, Map } from 'immutable';
import { TodoList } from '../TodoList';


describe('<TodoList/>', () => {
    test('should render TodoList', () => {
        const data = Map({
            todos: List([
                Map({ id: 1, isDone: false, text: 'todo 1', isDeleted: false }),
                Map({ id: 2, isDone: false, text: 'todo 2', isDeleted: false })
            ])
        });
        const wrapper = renderer.create(
            <TodoList
                currentData={data}
                addTodo={jest.fn()}
                updateTodo={jest.fn()}
                toggleTodo={jest.fn()}
                removeTodo={jest.fn()}
            />
        );
        const tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
