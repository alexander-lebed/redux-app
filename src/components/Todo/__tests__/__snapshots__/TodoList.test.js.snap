// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList/> should render TodoList 1`] = `
<div
  className="todo"
>
  <div
    className="row"
  >
    <div
      className="col-xs-3"
    />
    <div
      className="col-xs-6"
    >
      <input
        className="todo__entry"
        onKeyDown={[Function]}
        placeholder="Add todo"
        type="text"
      />
      <ul
        className="todo__list"
      >
        <li
          className="todo__item"
          onClick={[Function]}
        >
          <div>
            <span>
              todo 1
            </span>
            <div
              className="pull-right btn-toolbar"
              role="toolbar"
              style={
                Object {
                  "marginTop": 3,
                }
              }
            >
              <span
                className="glyphicon glyphicon-pencil"
                id="edit-todo"
                onClick={[Function]}
                style={
                  Object {
                    "color": "grey",
                    "marginRight": 15,
                  }
                }
              />
              <span
                className="glyphicon glyphicon-remove"
                id="remove-todo"
                onClick={[Function]}
                style={
                  Object {
                    "color": "grey",
                  }
                }
              />
            </div>
          </div>
        </li>
        <li
          className="todo__item"
          onClick={[Function]}
        >
          <div>
            <span>
              todo 2
            </span>
            <div
              className="pull-right btn-toolbar"
              role="toolbar"
              style={
                Object {
                  "marginTop": 3,
                }
              }
            >
              <span
                className="glyphicon glyphicon-pencil"
                id="edit-todo"
                onClick={[Function]}
                style={
                  Object {
                    "color": "grey",
                    "marginRight": 15,
                  }
                }
              />
              <span
                className="glyphicon glyphicon-remove"
                id="remove-todo"
                onClick={[Function]}
                style={
                  Object {
                    "color": "grey",
                  }
                }
              />
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      className="col-xs-3"
    />
  </div>
</div>
`;
